File: merge-sort.cpp
#include <vector>
#include <iostream>

using namespace std;

void merge(vector<int> &nums, int left, int middle, int right) {
  vector<int> leftArray(nums.begin() + left, nums.begin() + middle);
  vector<int> rightArray(nums.begin() + middle, nums.begin() + right + 1);

  int i = 0, j = 0, k = left;
  while(i < leftArray.size() && j < rightArray.size()){
    if (leftArray[i] <= rightArray[j]) {
      nums[k] = leftArray[i];
      i++;
    } else {
      nums[k] = rightArray[j];
      j++;
    }
    k++;
  }
  
  while(i < leftArray.size()){
    nums[k] = leftArray[i];
    i++;
    k++;
  }

  while(j < rightArray[j]){
    /* code */
    nums[k] = rightArray[j];
    j++;
    k++;
  }
}

void mergeSort(vector<int> &nums, int left, int right) {
  if (left < right) {
    int middle = left + (right - left) / 2;
    mergeSort(nums, left, middle);
    mergeSort(nums, middle, right);
    merge(nums, left, middle, right);
  }
  return;
}

void sort(vector<int> &nums) {
  mergeSort(nums, 0, nums.size() - 1);
}

int main() {
  vector<int> nums;
  nums.push_back(3);
  nums.push_back(2);
  sort(nums);
}

File: sort/merge-sort.js
function mergeSort(nums, left, right) {
  if (left < right) {
    const middle = left + Math.floor((right - left) / 2)
    mergeSort(left, middle)
    mergeSort(middle + 1, right)
    merge(nums, left, middle, right)
  }
}

function merge(nums, left, middle, right) {
  const leftArray = nums.slice(left, middle)
  const rightArray = nums.slice(middle, right + 1)
  let i = 0, j = 0, k = left
  while (i < leftArray.length && j < rightArray.length) {
    if (leftArray[i] < rightArray[j]) {
      nums[k] = leftArray[i]
      i++
    } else {
      nums[k] = rightArray[j]
      j++
    }
    k++
  }
  while (i < leftArray) {
    nums[k] = leftArray[i]
    i++
    k++
  }
  while (j < rightArray) {
    nums[k] = rightArray[j]
    j++
    k++
  }
}

const nums = [3, 2, 1]
mergeSort(nums, 0 , 2)
console.log(nums)